astroid/interpreter/_import/util.py

```diff
@@ -18,17 +18,18 @@ def is_namespace(modname: str) -> bool:
     # That's unacceptable here, so we fallback to _find_spec_from_path(), which does
     # not, but requires instead that each single parent ('astroid', 'nodes', etc.)
     # be specced from left to right.
-    processed_components = []
-    last_parent = None
-    for component in modname.split("."):
-        processed_components.append(component)
-        working_modname = ".".join(processed_components)
+    components = modname.split(".")
+    for i in range(1, len(components) + 1):
+        working_modname = ".".join(components[:i])
         try:
-            found_spec = _find_spec_from_path(working_modname, last_parent)
+            # Search under the highest package name
+            # Only relevant if package not already on sys.path
+            # See https://github.com/python/cpython/issues/89754 for reasoning
+            # Otherwise can raise bare KeyError: https://github.com/python/cpython/issues/93334
+            found_spec = _find_spec_from_path(working_modname, components[0])
```

Comment:
@jacobtylerwalls Just thinking about this again. We cached this function because of `importlib` being slow in https://github.com/PyCQA/astroid/pull/1536#discussion_r864481144. (My very first review comment there).

What if we try to get the spec of:
``a.b.c.d.e.f.g`` and then ``a.b.c.d.e.f.h``? We don't benefit from any of the caching right?

Reply:
True, but I don't know if we ever confirmed that this use of importlib was slow. It's slow if importlib is actually importing the parent modules (why we did the #1409 to revert) but the whole point of this approach was to not do that, so I'm not sure it's an issue.